### PR TITLE
refactor(uri-parser): support authentication expectations

### PR DESCRIPTION
**Pull request by mbroadst**
Adding support for the expectations put forth in the authentication spec, as well as the accompanied tests.

Fixes NODE-1689

### CODE REVIEW

File: lib/uri_parser.js [hunk 1]
```diff
@@ -296,6 +296,76 @@ function applyConnectionStringOption(obj, key, value, options) {
...
+      parsed.auth = Object.assign({}, parsed.auth, { db: '$external' });
+    }
+
+    if (authMechanism === 'MONGODB-X509') {
+      if (parsed.auth && parsed.auth.password != null) {
+        throw new MongoParseError(`Password not allowed for mechanism \`${authMechanism}\``);
+      }
+
+      if (authSource != null && authSource !== '$external') {
+        throw new MongoParseError(`Invalid auth source \`${authMechanism}\` specified`);
```
Comment (kvwalker) [inline lib/uri_parser.js]: Same note as above.

File: lib/uri_parser.js [hunk 2]
```diff
@@ -296,6 +296,76 @@ function applyConnectionStringOption(obj, key, value, options) {
...
+    if (
+      USERNAME_REQUIRED_MECHANISMS.has(authMechanism) &&
+      (!parsed.auth || parsed.auth.username == null)
+    ) {
+      throw new MongoParseError(`Username required for mechanism \`${authMechanism}\``);
+    }
+
+    if (authMechanism === 'GSSAPI') {
+      if (authSource != null && authSource !== '$external') {
+        throw new MongoParseError(`Invalid auth source \`${authMechanism}\` specified`);
```
Comment (kvwalker) [inline lib/uri_parser.js]: We could make this error message more clear, like `Invalid authSource ${authSource} for mechanism ${authMechanism} specified.`

File: lib/uri_parser.js [hunk 3]
```diff
@@ -296,6 +296,76 @@ function applyConnectionStringOption(obj, key, value, options) {
...
+  if (authMechanism != null) {
+    if (
+      USERNAME_REQUIRED_MECHANISMS.has(authMechanism) &&
+      (!parsed.auth || parsed.auth.username == null)
+    ) {
+      throw new MongoParseError(`Username required for mechanism \`${authMechanism}\``);
+    }
+
+    if (authMechanism === 'GSSAPI') {
+      if (authSource != null && authSource !== '$external') {
```
Comment (daprahamian) [inline lib/uri_parser.js]: Does this rule also apply to `X509`?

Comment (daprahamian) [reply lib/uri_parser.js]: NVM you did it below :)

File: lib/uri_parser.js [hunk 4]
```diff
@@ -296,6 +296,76 @@ function applyConnectionStringOption(obj, key, value, options) {
...
+  }
+
+  const options = parsed.options;
+  const authSource = options.authsource || options.authSource;
+  if (authSource != null) {
+    parsed.auth = Object.assign({}, parsed.auth, { db: authSource });
+  }
+
+  const authMechanism = options.authmechanism || options.authMechanism;
+  if (authMechanism != null) {
```
Comment (daprahamian) [inline lib/uri_parser.js]: Can we break up some of the sub-`if`s into sub-functions? This `if` seems really big

File: test/tests/unit/mongodb_srv_tests.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
+'use strict';
```
Comment (daprahamian) [inline test/tests/unit/mongodb_srv_tests.js]: Double checking: this is just renaming `mongodb_srv_spec_tests.js`?

Comment (mbroadst) [reply test/tests/unit/mongodb_srv_tests.js]: it does that, but also split the non-spec test involved into its own `describe` separate from the spec tests, and fixes that test not to use a SRV host that returns an `authSource`

File: lib/uri_parser.js (hunk 4 above)
Comment (mbroadst) [reply lib/uri_parser.js]: This is what I was referring to yesterday in our meeting. I think all of these, in a more elegantly designed system, would live in some sort of factory method on each of the sasl modules. As it stands, its a little long but its a literal parse and I'd prefer to actually reduce the function calls unless its bringing a great increase in readability. Do you feel strongly about this?

File: test/tests/unit/mongodb_srv_tests.js (hunk 5 above)
Comment (daprahamian) [reply test/tests/unit/mongodb_srv_tests.js]: nah, if we moved it to the authproviders later that would be fine. Can we file a ticket for it?

File: lib/uri_parser.js (hunk 4 above)
Comment (daprahamian) [reply lib/uri_parser.js]: Put this comment on the wrong thread: nah, if we moved it to the authproviders later that would be fine. Can we file a ticket for it?

Comment (mbroadst) [reply lib/uri_parser.js]: Done. [NODE-1691](https://jira.mongodb.org/browse/NODE-1691)

**Pull request merged**